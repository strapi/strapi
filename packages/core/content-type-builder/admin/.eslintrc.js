module.exports = {
  root: true,
  overrides: [
    {
      files: ['**/*.ts', '**/*.tsx'],
      extends: ['custom/front/typescript'],
    },
  ],
};
