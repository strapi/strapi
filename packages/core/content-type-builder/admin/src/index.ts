import { Layout } from '@strapi/icons';

import { PERMISSIONS } from './constants';
import { pluginId } from './pluginId';
import { reducers } from './reducers';
import { formsAPI } from './utils/formAPI';
import { prefixPluginTranslations } from './utils/prefixPluginTranslations';

import type { StrapiApp } from '@strapi/admin/strapi-admin';

// eslint-disable-next-line import/no-default-export
export default {
  register(app: StrapiApp) {
    app.addReducers(reducers);
    app.addMenuLink({
      to: `plugins/${pluginId}`,
      icon: Layout,
      intlLabel: {
        id: `${pluginId}.plugin.name`,
        defaultMessage: 'Content-Type Builder',
      },
      permissions: PERMISSIONS.main,
      Component: () => import('./pages/App'),
      position: 5,
    });

    app.registerPlugin({
      id: pluginId,
      name: pluginId,
      // Internal APIs exposed by the CTB for the other plugins to use
      apis: {
        forms: formsAPI,
      },
    });
  },
  bootstrap() {},
  async registerTrads({ locales }: { locales: string[] }) {
    const importedTrads = await Promise.all(
      locales.map((locale) => {
        return import(`./translations/${locale}.json`)
          .then(({ default: data }) => {
            return {
              data: prefixPluginTranslations(data, pluginId),
              locale,
            };
          })
          .catch(() => {
            return {
              data: {},
              locale,
            };
          });
      })
    );

    return Promise.resolve(importedTrads);
  },
};

export * from './exports';
