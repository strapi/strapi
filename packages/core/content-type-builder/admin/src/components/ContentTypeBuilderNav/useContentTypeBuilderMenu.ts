import { useState, MouseEvent } from 'react';

import { useTracking, useNotification } from '@strapi/admin/strapi-admin';
import { useCollator, useFilter } from '@strapi/design-system';
import isEqual from 'lodash/isEqual';
import { useIntl } from 'react-intl';

import { useDataManager } from '../../hooks/useDataManager';
import { useFormModalNavigation } from '../../hooks/useFormModalNavigation';
import { pluginId } from '../../pluginId';
import { getTrad } from '../../utils/getTrad';

import type { Internal } from '@strapi/types';

export const useContentTypeBuilderMenu = () => {
  const {
    components,
    componentsGroupedByCategory,
    contentTypes,
    isInDevelopmentMode,
    sortedContentTypesList,
    modifiedData,
    initialData,
  } = useDataManager();
  const { toggleNotification } = useNotification();
  const { formatMessage } = useIntl();
  const { trackUsage } = useTracking();
  const [searchValue, setSearchValue] = useState('');
  const { onOpenModalCreateSchema, onOpenModalEditCategory } = useFormModalNavigation();
  const { locale } = useIntl();

  const { startsWith } = useFilter(locale, {
    sensitivity: 'base',
  });

  const formatter = useCollator(locale, {
    sensitivity: 'base',
  });

  // TODO: Allow creating mutliple schemas in parallel
  const canOpenModalCreateCTorComponent =
    !Object.keys(contentTypes).some((ct) => contentTypes[ct].isTemporary === true) &&
    !Object.keys(components).some(
      (component) => components[component as Internal.UID.Component].isTemporary === true
    ) &&
    isEqual(modifiedData, initialData);

  const handleClickOpenModalCreateCollectionType = () => {
    if (canOpenModalCreateCTorComponent) {
      // TODO: Review tracking with product
      trackUsage(`willCreateContentType`);

      const nextState = {
        modalType: 'contentType',
        kind: 'collectionType',
        actionType: 'create',
        forTarget: 'contentType',
      };

      onOpenModalCreateSchema(nextState);
    } else {
      toggleNotificationCannotCreateSchema();
    }
  };

  const handleClickOpenModalCreateSingleType = () => {
    if (canOpenModalCreateCTorComponent) {
      trackUsage(`willCreateSingleType`);

      const nextState = {
        modalType: 'contentType',
        kind: 'singleType',
        actionType: 'create',
        forTarget: 'contentType',
      };

      onOpenModalCreateSchema(nextState);
    } else {
      toggleNotificationCannotCreateSchema();
    }
  };

  const handleClickOpenModalCreateComponent = () => {
    if (canOpenModalCreateCTorComponent) {
      trackUsage('willCreateComponent');

      const nextState = {
        modalType: 'component',
        kind: null,
        actionType: 'create',
        forTarget: 'component',
      };

      onOpenModalCreateSchema(nextState);
    } else {
      toggleNotificationCannotCreateSchema();
    }
  };

  const toggleNotificationCannotCreateSchema = () => {
    toggleNotification({
      type: 'info',
      message: formatMessage({
        id: getTrad('notification.info.creating.notSaved'),
        defaultMessage: 'Please save your work before creating a new collection type or component',
      }),
    });
  };

  const componentsData = Object.entries(componentsGroupedByCategory)
    .map(([category, components]) => ({
      name: category,
      title: category,
      isEditable: isInDevelopmentMode,
      // TODO: re-add functionality to edit category name
      onClickEdit(e: MouseEvent, data: any) {
        e.stopPropagation();

        if (canOpenModalCreateCTorComponent) {
          onOpenModalEditCategory(data.name);
        } else {
          toggleNotificationCannotCreateSchema();
        }
      },
      links: components
        .map((component) => ({
          name: component.uid,
          to: `/plugins/${pluginId}/component-categories/${category}/${component.uid}`,
          title: component.schema.displayName,
        }))
        .sort((a, b) => formatter.compare(a.title, b.title)),
    }))
    .sort((a, b) => formatter.compare(a.title, b.title));

  const displayedContentTypes = sortedContentTypesList.filter((obj) => obj.visible);

  const data = [
    {
      name: 'models',
      title: {
        id: `${getTrad('menu.section.models.name')}`,
        defaultMessage: 'Collection Types',
      },
      customLink: isInDevelopmentMode && {
        id: `${getTrad('button.model.create')}`,
        defaultMessage: 'Create new collection type',
        onClick: handleClickOpenModalCreateCollectionType,
      },
      links: displayedContentTypes.filter((contentType) => contentType.kind === 'collectionType'),
    },
    {
      name: 'singleTypes',
      title: {
        id: `${getTrad('menu.section.single-types.name')}`,
        defaultMessage: 'Single Types',
      },
      customLink: isInDevelopmentMode && {
        id: `${getTrad('button.single-types.create')}`,
        defaultMessage: 'Create new single type',
        onClick: handleClickOpenModalCreateSingleType,
      },
      links: displayedContentTypes.filter((singleType) => singleType.kind === 'singleType'),
    },
    {
      name: 'components',
      title: {
        id: `${getTrad('menu.section.components.name')}`,
        defaultMessage: 'Components',
      },
      customLink: isInDevelopmentMode && {
        id: `${getTrad('button.component.create')}`,
        defaultMessage: 'Create a new component',
        onClick: handleClickOpenModalCreateComponent,
      },
      links: componentsData,
    },
  ].map((section) => {
    const hasChild = section.links.some((l) => Array.isArray(l.links));

    if (hasChild) {
      let filteredLinksCount = 0;

      return {
        ...section,
        links: section.links
          .map((link) => {
            const filteredLinks = link.links.filter((link: any) =>
              startsWith(link.title, searchValue)
            );

            if (filteredLinks.length === 0) {
              return null;
            }

            filteredLinksCount += filteredLinks.length;

            return {
              ...link,
              links: filteredLinks.sort((a: any, b: any) => formatter.compare(a.title, b.title)),
            };
          })
          .filter(Boolean),
        linksCount: filteredLinksCount,
      };
    }

    const filteredLinks = section.links
      .filter((link) => startsWith(link.title, searchValue))
      .sort((a, b) => formatter.compare(a.title, b.title));

    return {
      ...section,
      links: filteredLinks,
      linksCount: filteredLinks.length,
    };
  });

  return {
    menu: data,
    search: {
      value: searchValue,
      onChange: setSearchValue,
      clear: () => setSearchValue(''),
    },
  };
};
