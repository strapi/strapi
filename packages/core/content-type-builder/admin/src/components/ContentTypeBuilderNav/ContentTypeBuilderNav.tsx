import { Fragment } from 'react';

import {
  Box,
  TextButton,
  SubNav,
  SubNavHeader,
  SubNavLink,
  SubNavLinkSection,
  SubNavSection,
  SubNavSections,
} from '@strapi/design-system';
import { Plus } from '@strapi/icons';
import upperFirst from 'lodash/upperFirst';
import { useIntl } from 'react-intl';
import { NavLink } from 'react-router-dom';
import { styled } from 'styled-components';

import { getTrad } from '../../utils/getTrad';

import { useContentTypeBuilderMenu } from './useContentTypeBuilderMenu';

const SubNavLinkCustom = styled(SubNavLink)`
  div {
    width: inherit;
    span:nth-child(2) {
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      width: inherit;
    }
  }
`;

export const ContentTypeBuilderNav = () => {
  const { menu, search } = useContentTypeBuilderMenu();
  const { formatMessage } = useIntl();

  const pluginName = formatMessage({
    id: getTrad('plugin.name'),
    defaultMessage: 'Content-Type Builder',
  });

  return (
    <SubNav aria-label={pluginName}>
      <SubNavHeader
        searchable
        value={search.value}
        onClear={() => search.clear()}
        onChange={(e) => search.onChange(e.target.value)}
        label={pluginName}
        searchLabel={formatMessage({
          id: 'global.search',
          defaultMessage: 'Search',
        })}
      />
      <SubNavSections>
        {menu.map((section) => (
          <Fragment key={section.name}>
            <SubNavSection
              label={formatMessage({
                id: section.title.id,
                defaultMessage: section.title.defaultMessage,
              })}
              collapsable
              badgeLabel={section.linksCount.toString()}
            >
              {section.links.map((link) => {
                if (link.links) {
                  return (
                    <SubNavLinkSection key={link.name} label={upperFirst(link.title)}>
                      {link.links.map((subLink: any) => (
                        <SubNavLink
                          tag={NavLink}
                          to={subLink.to}
                          active={subLink.active}
                          key={subLink.name}
                          isSubSectionChild
                        >
                          {upperFirst(
                            formatMessage({ id: subLink.name, defaultMessage: subLink.title })
                          )}
                        </SubNavLink>
                      ))}
                    </SubNavLinkSection>
                  );
                }

                return (
                  <SubNavLinkCustom
                    tag={NavLink}
                    to={link.to}
                    active={link.active}
                    key={link.name}
                    width="100%"
                  >
                    {upperFirst(formatMessage({ id: link.name, defaultMessage: link.title }))}
                  </SubNavLinkCustom>
                );
              })}
            </SubNavSection>
            {section.customLink && (
              <Box paddingLeft={7}>
                <TextButton
                  onClick={section.customLink.onClick}
                  startIcon={<Plus width="0.8rem" height="0.8rem" />}
                  marginTop={2}
                  cursor="pointer"
                >
                  {formatMessage({
                    id: section.customLink.id,
                    defaultMessage: section.customLink.defaultMessage,
                  })}
                </TextButton>
              </Box>
            )}
          </Fragment>
        ))}
      </SubNavSections>
    </SubNav>
  );
};
