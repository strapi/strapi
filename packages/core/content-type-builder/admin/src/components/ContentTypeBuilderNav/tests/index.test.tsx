/* eslint-disable check-file/filename-naming-convention */
import { Layouts } from '@strapi/admin/strapi-admin';
import { DesignSystemProvider } from '@strapi/design-system';
import { render } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';

import { ContentTypeBuilderNav } from '../ContentTypeBuilderNav';

import { mockData } from './mockData';

jest.mock('../useContentTypeBuilderMenu.ts', () => {
  return {
    useContentTypeBuilderMenu: jest.fn(() => ({
      menu: mockData,
      search: {
        value: '',
        onChange() {},
      },
    })),
  };
});

const makeApp = () => {
  return (
    <IntlProvider messages={{}} defaultLocale="en" textComponent="span" locale="en">
      <DesignSystemProvider>
        <MemoryRouter>
          <Layouts.Root sideNav={<ContentTypeBuilderNav />}>
            <div />
          </Layouts.Root>
        </MemoryRouter>
      </DesignSystemProvider>
    </IntlProvider>
  );
};

describe('<ContentTypeBuilderNav />', () => {
  it('renders and matches the snapshot', () => {
    const App = makeApp();
    const { container } = render(App);

    expect(container).toMatchSnapshot();
  });
});
