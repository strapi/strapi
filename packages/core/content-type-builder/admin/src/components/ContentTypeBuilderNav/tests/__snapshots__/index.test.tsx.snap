// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentTypeBuilderNav /> renders and matches the snapshot 1`] = `
.c5 {
  background: #eaeaef;
}

.c7 {
  padding-block-start: 20px;
  padding-inline-end: 20px;
  padding-block-end: 20px;
  padding-inline-start: 20px;
}

.c10 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  width: 100%;
  border-radius: 4px;
  cursor: pointer;
  flex: 1;
}

.c14 {
  padding-block-start: 4px;
  padding-inline-start: 7px;
  padding-inline-end: 7px;
  border-radius: 4px;
  cursor: pointer;
}

.c22 {
  border-radius: 4px;
  display: inline-flex;
  cursor: pointer;
}

.c25 {
  padding-block-end: 16px;
}

.c28 {
  padding-inline-start: 20px;
  padding-inline-end: 20px;
}

.c29 {
  position: relative;
}

.c33 {
  margin-inline-start: 8px;
  margin-inline-end: 8px;
}

.c37 {
  padding-inline-start: 12px;
  padding-inline-end: 12px;
  width: 100%;
  border-radius: 4px;
}

.c38 {
  width: 100%;
}

.c40 {
  width: 100%;
  overflow: hidden;
}

.c42 {
  width: 24px;
  text-align: center;
}

.c44 {
  padding-inline-start: 8px;
}

.c47 {
  padding-block-end: 56px;
}

.c8 {
  gap: 12px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c9 {
  gap: 8px;
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c11 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c17 {
  gap: 8px;
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c20 {
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c23 {
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: flex;
}

.c26 {
  gap: 20px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c27 {
  gap: 8px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c30 {
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c34 {
  gap: 2px;
  align-items: stretch;
  flex-direction: column;
  display: flex;
}

.c39 {
  gap: 4px;
  align-items: center;
  justify-content: space-between;
  flex-direction: row;
  display: flex;
}

.c4 {
  font-weight: 600;
  font-size: 1.8rem;
  line-height: 1.22;
  color: currentcolor;
}

.c13 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c31 {
  font-weight: 600;
  font-size: 1.1rem;
  line-height: 1.45;
  text-transform: uppercase;
  color: #666687;
}

.c41 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: currentcolor;
  line-height: 32px;
}

.c45 {
  font-size: 1.4rem;
  line-height: 1.43;
  color: #32324d;
  font-weight: 500;
}

.c16 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c12 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #4945ff;
  background: #4945ff;
  color: #ffffff;
}

.c12:hover {
  border: 1px solid #7b79ff;
  background: #7b79ff;
}

.c12:active {
  border: 1px solid #4945ff;
  background: #4945ff;
}

.c12[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c15 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #dcdce4;
  background: #ffffff;
  color: #32324d;
}

.c15:hover {
  background-color: #f6f6f9;
}

.c15:active {
  background-color: #eaeaef;
}

.c15[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c24 {
  text-decoration: none;
  padding-block: 0.7rem;
  padding-inline: 0.7rem;
  border: 1px solid transparent;
  background: transparent;
  color: #32324d;
}

.c24 svg {
  fill: #8e8ea9;
}

.c24:hover {
  background-color: #f6f6f9;
}

.c24:active {
  background-color: #eaeaef;
}

.c24[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c32 {
  text-decoration: none;
  padding-block: 0.25rem;
  padding-inline: 0.25rem;
  border: 1px solid transparent;
  background: transparent;
  color: #32324d;
}

.c32 svg {
  fill: #8e8ea9;
}

.c32:hover {
  background-color: #f6f6f9;
}

.c32:active {
  background-color: #eaeaef;
}

.c32[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

.c19 {
  border: none;
  border-radius: 4px;
  color: #32324d;
  font-weight: 400;
  font-size: 1.4rem;
  line-height: 2.2rem;
  display: block;
  width: 100%;
  background: inherit;
  padding-inline-start: 0;
  padding-inline-end: 0;
  padding-block: 4px;
}

.c19::placeholder {
  color: #666687;
  opacity: 1;
}

.c19[aria-disabled='true'] {
  color: inherit;
}

.c19:focus {
  outline: none;
  box-shadow: none;
}

.c21 {
  position: absolute;
  right: 16px;
  top: 50%;
  transform: translateY(-50%);
}

.c18 {
  border: 1px solid #dcdce4;
  border-radius: 4px;
  background: #ffffff;
  padding-inline-start: 16px;
  position: relative;
  outline: none;
  box-shadow: none;
  transition-property: border-color,box-shadow,fill;
  transition-duration: 0.2s;
}

.c18:focus-within {
  border: 1px solid #4945ff;
  box-shadow: #4945ff 0px 0px 0px 2px;
}

.c6 {
  height: 1px;
  border: none;
  flex-shrink: 0;
}

.c1 {
  width: 23.2rem;
  background: #f6f6f9;
  position: sticky;
  top: 0;
  height: 100vh;
  overflow-y: auto;
  border-right: 1px solid #dcdce4;
  z-index: 1;
}

.c0 {
  display: grid;
  grid-template-columns: auto 1fr;
}

.c48 {
  overflow-x: hidden;
}

.c2 {
  background-color: #ffffff;
  border-right: 1px solid #eaeaef;
  scrollbar-width: none;
  -ms-overflow-style: none;
}

.c2::-webkit-scrollbar {
  display: none;
}

.c36 {
  display: flex;
  align-items: center;
  justify-content: space-between;
  text-decoration: none;
  height: 32px;
  color: #32324d;
}

.c36.active>div {
  background-color: #f0f0ff;
  color: #271fe0;
  font-weight: 500;
}

.c36:hover.active>div {
  background-color: #f0f0ff;
}

.c36:hover>div {
  background-color: #f6f6f9;
}

.c36:focus-visible {
  outline-offset: -2px;
}

.c3 {
  height: 56px;
  display: flex;
  align-items: center;
  padding-left: 20px;
}

.c43 {
  cursor: pointer;
  width: 100%;
  border: none;
  padding: 0;
  background: transparent;
  display: flex;
  align-items: center;
  height: 32px;
  border-radius: 4px;
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
  padding-bottom: 8px;
}

.c43:hover {
  background-color: #f6f6f9;
}

.c46 .c35>div {
  padding-left: 36px;
}

@media (prefers-reduced-motion: no-preference) {
  .c12 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c15 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c24 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

@media (prefers-reduced-motion: no-preference) {
  .c32 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

<div>
  <div
    class="c0"
  >
    <nav
      aria-label="Content-Type Builder"
      class="c1 c2"
    >
      <div
        class="c3"
      >
        <h2
          class="c4"
        >
          Content-Type Builder
        </h2>
      </div>
      <div
        class="c5 c6"
        data-orientation="horizontal"
        role="separator"
      />
      <div
        class="c7 c8"
      >
        <div
          class="c9"
        >
          <button
            aria-disabled="false"
            class="c10 c11 c12"
            type="submit"
          >
            <span
              class="c13"
            >
              Save
            </span>
          </button>
          <button
            aria-disabled="false"
            aria-expanded="false"
            aria-haspopup="menu"
            class="c14 c11 c15"
            data-state="closed"
            id="radix-:r0:"
            type="button"
          >
            <span
              class="c13"
            >
              <svg
                aria-hidden="true"
                fill="#8e8ea9"
                focusable="false"
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M18 16a2 2 0 1 1-4 0 2 2 0 0 1 4 0M7.5 14a2 2 0 1 0 0 4 2 2 0 0 0 0-4m17 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4"
                />
              </svg>
              <span
                class="c16"
              >
                More actions
              </span>
            </span>
          </button>
        </div>
        <div
          class="c17 c18"
        >
          <svg
            fill="#8e8ea9"
            height="16"
            viewBox="0 0 32 32"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M29.061 26.939 23.125 21A11.515 11.515 0 1 0 21 23.125l5.941 5.942a1.503 1.503 0 0 0 2.125-2.125zM5.5 14a8.5 8.5 0 1 1 8.5 8.5A8.51 8.51 0 0 1 5.5 14"
            />
          </svg>
          <input
            aria-disabled="false"
            aria-label="Search"
            class="c19"
            placeholder="Search"
            style="padding-right: 24px;"
            value=""
          />
          <div
            class="c20 c21"
          >
            <button
              aria-disabled="false"
              class="c22 c23 c24"
              data-state="closed"
              style="padding: 0px;"
              type="button"
            >
              <svg
                aria-hidden="true"
                fill="currentColor"
                focusable="false"
                height="16"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M26.061 23.939a1.503 1.503 0 0 1-2.125 2.125L16 18.125l-7.939 7.936a1.503 1.503 0 1 1-2.125-2.125L13.875 16 5.939 8.061a1.503 1.503 0 1 1 2.125-2.125L16 13.875l7.939-7.94a1.502 1.502 0 1 1 2.125 2.125L18.125 16z"
                />
              </svg>
              <span
                class="c16"
              >
                clear
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="c25"
      >
        <ol
          class="c26"
        >
          <li>
            <div
              class="c27"
            >
              <div
                class="c28"
              >
                <div
                  class="c29 c30"
                >
                  <div
                    class="c20"
                  >
                    <div
                      class=""
                    >
                      <span
                        class="c31"
                      >
                        Collection Types
                      </span>
                    </div>
                  </div>
                  <button
                    aria-disabled="false"
                    class="c22 c23 c32"
                    data-state="closed"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5h-9.5V27a1.5 1.5 0 1 1-3 0v-9.5H5a1.5 1.5 0 1 1 0-3h9.5V5a1.5 1.5 0 1 1 3 0v9.5H27a1.5 1.5 0 0 1 1.5 1.5"
                      />
                    </svg>
                    <span
                      class="c16"
                    >
                      Create new collection type
                    </span>
                  </button>
                </div>
              </div>
              <ol
                class="c33 c34"
                id=":r3:"
              >
                <li>
                  <a
                    class="c35 c36"
                    href="/plugins/content-type-builder/content-types/application::address.address"
                  >
                    <div
                      class="c37"
                    >
                      <div
                        class="c38 c39"
                      >
                        <div
                          class="c40 c41"
                          style="text-overflow: ellipsis; white-space: nowrap;"
                        >
                          address
                        </div>
                        <div
                          class="c9"
                        >
                          <span
                            class="c42"
                          />
                        </div>
                      </div>
                    </div>
                  </a>
                </li>
                <li>
                  <a
                    class="c35 c36"
                    href="/plugins/content-type-builder/content-types/application::category.category"
                  >
                    <div
                      class="c37"
                    >
                      <div
                        class="c38 c39"
                      >
                        <div
                          class="c40 c41"
                          style="text-overflow: ellipsis; white-space: nowrap;"
                        >
                          category
                        </div>
                        <div
                          class="c9"
                        >
                          <span
                            class="c42"
                          />
                        </div>
                      </div>
                    </div>
                  </a>
                </li>
              </ol>
            </div>
          </li>
          <li>
            <div
              class="c27"
            >
              <div
                class="c28"
              >
                <div
                  class="c29 c30"
                >
                  <div
                    class="c20"
                  >
                    <div
                      class=""
                    >
                      <span
                        class="c31"
                      >
                        Single Types
                      </span>
                    </div>
                  </div>
                  <button
                    aria-disabled="false"
                    class="c22 c23 c32"
                    data-state="closed"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5h-9.5V27a1.5 1.5 0 1 1-3 0v-9.5H5a1.5 1.5 0 1 1 0-3h9.5V5a1.5 1.5 0 1 1 3 0v9.5H27a1.5 1.5 0 0 1 1.5 1.5"
                      />
                    </svg>
                    <span
                      class="c16"
                    >
                      Create new single type
                    </span>
                  </button>
                </div>
              </div>
              <ol
                class="c33 c34"
                id=":r5:"
              >
                <li>
                  <a
                    class="c35 c36"
                    href="/plugins/content-type-builder/content-types/application::homepage.homepage"
                  >
                    <div
                      class="c37"
                    >
                      <div
                        class="c38 c39"
                      >
                        <div
                          class="c40 c41"
                          style="text-overflow: ellipsis; white-space: nowrap;"
                        >
                          Homepage
                        </div>
                        <div
                          class="c9"
                        >
                          <span
                            class="c42"
                          />
                        </div>
                      </div>
                    </div>
                  </a>
                </li>
              </ol>
            </div>
          </li>
          <li>
            <div
              class="c27"
            >
              <div
                class="c28"
              >
                <div
                  class="c29 c30"
                >
                  <div
                    class="c20"
                  >
                    <div
                      class=""
                    >
                      <span
                        class="c31"
                      >
                        Components
                      </span>
                    </div>
                  </div>
                  <button
                    aria-disabled="false"
                    class="c22 c23 c32"
                    data-state="closed"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      focusable="false"
                      height="16"
                      viewBox="0 0 32 32"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M28.5 16a1.5 1.5 0 0 1-1.5 1.5h-9.5V27a1.5 1.5 0 1 1-3 0v-9.5H5a1.5 1.5 0 1 1 0-3h9.5V5a1.5 1.5 0 1 1 3 0v9.5H27a1.5 1.5 0 0 1 1.5 1.5"
                      />
                    </svg>
                    <span
                      class="c16"
                    >
                      Create a new component
                    </span>
                  </button>
                </div>
              </div>
              <ol
                class="c33 c34"
                id=":r7:"
              >
                <li>
                  <div
                    class=""
                  >
                    <div
                      class="c30"
                    >
                      <button
                        aria-controls=":r9:"
                        aria-expanded="true"
                        class="c43"
                      >
                        <svg
                          aria-hidden="true"
                          fill="#8e8ea9"
                          height="16"
                          style="transform: rotate(0deg); transition: transform 0.5s;"
                          viewBox="0 0 32 32"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m27.061 13.061-10 10a1.503 1.503 0 0 1-2.125 0l-10-10a1.503 1.503 0 1 1 2.125-2.125L16 19.875l8.939-8.94a1.502 1.502 0 1 1 2.125 2.125z"
                          />
                        </svg>
                        <div
                          class="c44"
                        >
                          <span
                            class="c45"
                          >
                            basic
                          </span>
                        </div>
                      </button>
                    </div>
                    <ul
                      class="c34"
                      id=":r9:"
                      style="max-height: 1000px; overflow: hidden; transition: max-height 1s ease-in-out;"
                    >
                      <li
                        class="c46"
                      >
                        <a
                          class="c35 c36"
                          href="/plugins/content-type-builder/component-categories/basic/basic.simple"
                        >
                          <div
                            class="c37"
                          >
                            <div
                              class="c38 c39"
                            >
                              <div
                                class="c40 c41"
                                style="text-overflow: ellipsis; white-space: nowrap;"
                              >
                                simple
                              </div>
                              <div
                                class="c9"
                              >
                                <span
                                  class="c42"
                                />
                              </div>
                            </div>
                          </div>
                        </a>
                      </li>
                    </ul>
                  </div>
                </li>
                <li>
                  <div
                    class=""
                  >
                    <div
                      class="c30"
                    >
                      <button
                        aria-controls=":ra:"
                        aria-expanded="true"
                        class="c43"
                      >
                        <svg
                          aria-hidden="true"
                          fill="#8e8ea9"
                          height="16"
                          style="transform: rotate(0deg); transition: transform 0.5s;"
                          viewBox="0 0 32 32"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="m27.061 13.061-10 10a1.503 1.503 0 0 1-2.125 0l-10-10a1.503 1.503 0 1 1 2.125-2.125L16 19.875l8.939-8.94a1.502 1.502 0 1 1 2.125 2.125z"
                          />
                        </svg>
                        <div
                          class="c44"
                        >
                          <span
                            class="c45"
                          >
                            default
                          </span>
                        </div>
                      </button>
                    </div>
                    <ul
                      class="c34"
                      id=":ra:"
                      style="max-height: 1000px; overflow: hidden; transition: max-height 1s ease-in-out;"
                    >
                      <li
                        class="c46"
                      >
                        <a
                          class="c35 c36"
                          href="/plugins/content-type-builder/component-categories/default/default.closingperiod"
                        >
                          <div
                            class="c37"
                          >
                            <div
                              class="c38 c39"
                            >
                              <div
                                class="c40 c41"
                                style="text-overflow: ellipsis; white-space: nowrap;"
                              >
                                closingperiod
                              </div>
                              <div
                                class="c9"
                              >
                                <span
                                  class="c42"
                                />
                              </div>
                            </div>
                          </div>
                        </a>
                      </li>
                      <li
                        class="c46"
                      >
                        <a
                          class="c35 c36"
                          href="/plugins/content-type-builder/component-categories/default/default.dish"
                        >
                          <div
                            class="c37"
                          >
                            <div
                              class="c38 c39"
                            >
                              <div
                                class="c40 c41"
                                style="text-overflow: ellipsis; white-space: nowrap;"
                              >
                                dish
                              </div>
                              <div
                                class="c9"
                              >
                                <span
                                  class="c42"
                                />
                              </div>
                            </div>
                          </div>
                        </a>
                      </li>
                    </ul>
                  </div>
                </li>
              </ol>
            </div>
          </li>
        </ol>
      </div>
    </nav>
    <div
      class="c47 c48"
    >
      <div />
    </div>
  </div>
  <span
    class="c16"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
